src/gpl/src/mbff.cpp

```diff
@@ -43,58 +43,277 @@
+  const bool yes = (iterm->getIoType() == odb::dbIoType::INPUT);
+  return (yes & !exclude);
+}
+
+bool MBFF::IsQPin(odb::dbITerm* iterm)
+{
+  const bool exclude = (IsClockPin(iterm) || IsSupplyPin(iterm));
+  const bool yes = (iterm->getIoType() == odb::dbIoType::OUTPUT);
+  return (yes & !exclude);
+}
+
+int MBFF::GetNumD(odb::dbInst* inst)
+{
+  int num_d = 0;
+  for (auto iterm : inst->getITerms()) {
+    num_d += IsDPin(iterm);
+  }
+  return num_d;
+}
+
+int MBFF::GetNumQ(odb::dbInst* inst)
+{
+  int num_q = 0;
+  for (auto iterm : inst->getITerms()) {
+    num_q += IsQPin(iterm);
+  }
+  return num_q;
+}
+
+int MBFF::GetBitIdx(int bit_cnt)
+{
+  for (int i = 0; i < num_sizes_; i++) {
+    if ((1 << i) == bit_cnt) {
+      return i;
+    }
+  }
+  log_->error(utl::GPL, 122, "{} is not in 2^[0,{}]", bit_cnt, num_sizes_);
+}
 
 int MBFF::GetBitCnt(int bit_idx)
 {
-  if (bit_idx == 0) {
-    return 1;
+  return (1 << bit_idx);
+}
+
+int MBFF::GetRows(int slot_cnt, int setting)
+{
+  int idx = GetBitIdx(slot_cnt);
+  int width = int(multiplier_ * tray_width_[setting][idx]);
+  int height = int(multiplier_
+                   * (tray_area_[setting][idx] / tray_width_[setting][idx]));
+  return (height / std::gcd(width, height));
+}
+
+double MBFF::GetDist(const Point& a, const Point& b)
+{
+  return (abs(a.x - b.x) + abs(a.y - b.y));
+}
+
+bool MBFF::IsValidTray(odb::dbInst* tray)
+{
+  sta::Cell* cell = network_->dbToSta(tray->getMaster());
+  if (cell == nullptr) {
+    return false;
+  }
+  sta::LibertyCell* lib_cell = network_->libertyCell(cell);
+  if (lib_cell == nullptr) {
+    return false;
+  }
+  sta::LibertyCellPortIterator port_itr(lib_cell);
+  if (port_itr.hasNext() == false) {
+    return false;
   }
-  return (1 << (bit_idx + 1));
+  int slot_cnt = GetNumD(tray);
+  if (slot_cnt <= 1) {
+    return false;
+  }
+  return true;
 }
 
-int MBFF::GetRows(int slot_cnt)
+std::map<sta::LibertyPort*, std::pair<sta::LibertyPort*, sta::LibertyPort*>>
+MBFF::GetPinMapping(odb::dbInst* tray)
 {
-  if (slot_cnt == 1 || slot_cnt == 2 || slot_cnt == 4) {
-    return 1;
+  sta::Cell* cell = network_->dbToSta(tray->getMaster());
+  sta::LibertyCell* lib_cell = network_->libertyCell(cell);
+  sta::LibertyCellPortIterator port_itr(lib_cell);
+
+  std::vector<sta::LibertyPort*> d_pins;
+  std::vector<sta::LibertyPort*> q_pins;
+  std::vector<sta::LibertyPort*> qn_pins;
+
+  while (port_itr.hasNext()) {
+    sta::LibertyPort* port = port_itr.next();
+    if (network_->staToDb(port) == nullptr) {
+      continue;
+    }
+    if (port->isClock()) {
+      continue;
+    }
+    if (port->direction()->isInput()) {
+      d_pins.push_back(port);
+    }
+    if (port->direction()->isOutput()) {
+      if (q_pins.size()) {
+        if (sta::FuncExpr::equiv(q_pins.back()->function(), port->function())) {
+          q_pins.push_back(port);
+        } else {
+          qn_pins.push_back(port);
+        }
+      } else {
+        q_pins.push_back(port);
```

Comment:
Why assuming the first pin is Q and not QN?

Reply:
It ends up not mattering since we don't actually care about whether it is Q or QN.
As long as we group pins by different functions, the logic works.

Reply:
you can't group a d/clk/q with a d/clk/qn

Reply:
This won't happen because my program checks that q and qn have the same function.
In your given given, q and qn have different functions

Reply:
Maybe it makes more sense to think of the vectors q and qn has a collection of ports that have the same function.
If a "Q" port gets placed into the vector qn, then the "QN" port has to be placed into q since we check for the function equivalence.

When we cluster flops and modify the netlist, if I want to disconnect the FF's "QN" port and connect the tray's "QN" port, I check the vector q.
Since the port's in q have the same function as our "QN" port, we make a valid connection.